### ISSUE TITLE
Introduce implicit standalone mode

### ISSUE THREAD

**Issue by warpech** [Starcounter/Starcounter#3176]
Initially discussed here: https://github.com/StarcounterSamples/Barcodes2/issues/1

I have created an implementation of the implicit standalone mode. 

When a viewmodel extends `Partial` class instead of `Page`, then it gets this feature: when a response is sent to the browser, and it does not contain a `<!doctype>` (meaning it's a partial) it gets automatically wrapped with main HTML (that loads Puppet etc).

The intention is to deprecate `Page` and start using `Partial` for all sample apps, because it removes the boilerplate to make a standalone app.

Source https://github.com/StarcounterSamples/Partial/blob/master/Partial.cs

I will put it to Starcounter/Starcounter for further experiments in the apps.

**Comment by per-samuelsson** [comment]
@warpech 

You pushed it to the "master" branch. That one's not active. Can you push it to "develop"? Thanks.

**Comment by per-samuelsson** [comment]
> You pushed it to the "master" branch. 

No, sorry, I was wrong. Looked in samples (followed your link).

**Comment by per-samuelsson** [comment]
Ah, now I saw what was wrong - you pushed the file, but didn't add it to the project. So it's never built. I'll do that locally, and I'll push it if it works to build it.

**Comment by miyconst** [comment]
I have included the file into solution, replaced `ASCII` with `UTF8` encoding, and made all partials be statefull.

**Comment by miyconst** [comment]
I have two suggestions:

Add 

```html
<!-- This is default Starcounter template, you can change it by returning full html page with <!DOCTYPE> on the first line -->
```
to the default template.

Add 
```cs
if (this.Data == null) {
    this.OnData();
}
```

In the beginning of `AsMimeType`, so the `OnData` method always executed, even if no data assigned.

**Comment by warpech** [comment]
> Add `<!-- comment -->` to the default template.

Good idea, but let's make the comment shorter because it has a chance to finish in production (if Launcher also relies on implicit standalone mode).

> In the beginning of AsMimeType, so the OnData method always executed, even if no data assigned.

I think this needs to be consistent with whatever the `Json` class is doing. Perhaps you want to mitigate the fact that constructors can't be used in `.json.cs`? Maybe @chrhol could add some `OnInit` method to `Json` so solve that.

**Comment by miyconst** [comment]
Agree on both. I am fine with any text with would have same/similar meaning, and the `OnInit` method would really do the job.

**Comment by per-samuelsson** [comment]
> Maybe @chrhol could add some `OnInit` method to `Json` so solve that.

I noticed @jwillmer had introduced a specifc interface, like `IPageInitialisation`, that he applied to all of his pages, supporting generic initialization like that. Thats at least one indication this might be a good idea.

**Comment by warpech** [comment]
This is something we do in most of the apps somehow:

- in KitchenSink, we assign [`Data = null`](https://github.com/StarcounterSamples/KitchenSink/blob/662c2cb8cce97e59ad91101502f84eaaf72b5ac4/src/KitchenSink/Program.cs#L105) which triggers [`OnData`](https://github.com/StarcounterSamples/KitchenSink/blob/662c2cb8cce97e59ad91101502f84eaaf72b5ac4/src/KitchenSink/DatepickerPage.json.cs#L6)
- in Products, there is a [RefreshProducts](https://github.com/StarcounterSamples/Products/blob/990ba5e1749297c7a6fc060e70d11c186651bece/src/Products/ViewModels/ProductListPage.json.cs#L9) method called [after constructing](https://github.com/StarcounterSamples/Products/blob/990ba5e1749297c7a6fc060e70d11c186651bece/src/Products/Api/MainHandlers.cs#L85) of the object. Most of sample apps have a variation of this pattern.

**Comment by per-samuelsson** [comment]
Some initial feedback on this:

1. I've started using them in my super-slim PIM, and the basics work fine as far as I can tell - now I can run the PIM standalone, just doing `localhost:8080/pim` to get a nice page up. :+1: 
2. For some reason, a partial referenced from my main page does not render, and looking at the JSON in the debug aid shows me just 3 out of like 7-8 properties.
3. When I click a link in my main rendering, it doesn't switch to the linked resource (which probably is because of [2]).

I'm currently trying to understand whats causing 2+3, but still to no prevail. If you want to take a look @warpech, just clone and run https://github.com/per-samuelsson/PIM, open `localhost:8080/pim` and click one of the links.

**Comment by per-samuelsson** [comment]
Oh, and more about 2+3 from above:

* The [2] partial is `IBound`, while [1] is not (and [1] works). Can that be a problem?
* One of the properties not being returned is "Html", which I guess is why it doesn't render. :smile:

**Comment by per-samuelsson** [comment]
It seem to be related to the Mime type. For the ones that doesn't work, this statement

```C#
mimetype = req.PreferredMimeType;
```

return `MimeType.Unspecified`, which I guess is not what we want? Alternatively, I've modelled this wrong.

**Comment by warpech** [comment]
I see more than one difference between your app and StarcounterSamples, but let's take it one by one. 

The most importantly, I wonder why it doesn't work when I access this page by a full page load:  `http://10.211.55.4:8080/pim/types/6`. 

It seems that the JSON returned to the client is not the root of the partial. See the screenshot from the debug aid:

<img width="1438" alt="screen shot 2015-12-14 at 10 19 09" src="https://cloud.githubusercontent.com/assets/566463/11777153/a6d5182c-a24c-11e5-8e94-e3508795ede8.png">

Any idea why's that?

**Comment by warpech** [comment]
The above problem get's fixed when I remove this declaration from Partial.cs:

```
            if (Session.Current == null) {
                this.Session = new Session(SessionOptions.PatchVersioning);
            }
```

So it must be related to the implicit session. @miyconst it seems that you will need to tweak your idea

**Comment by warpech** [comment]
The problem is that the first `Partial` that creates the view model becomes the `Session.PublicViewModel`.

@chrhol do you remember discussing this? Wasn't the partial that is returned by `Handle.GET` supposed to become the `PublicViewModel`?

**Comment by warpech** [comment]
Until @chrhol can propose a fix, Problem 1 can be worked around by changing the line

```
partial class ProductTypeParent : Partial, IBound<ProductType> {
``` 

to 

```
partial class ProductTypeParent : Json, IBound<ProductType> {
```

This brings us to another:

**Problem 2**

When going to `http://localhost:8080/pim/`, then clicking on a link to `http://localhost:8080/pim/type/6`, the root properties of the view-model should be changed. Currently it is not the case. @chrhol any input on this one?

<img width="1431" alt="screen shot 2015-12-14 at 10 54 01" src="https://cloud.githubusercontent.com/assets/566463/11777835/05d579e4-a251-11e5-96f0-a119c4c99a2e.png">

Workaround for this one is to use a wrapper object (e.g. `master.CurrentPage = new ProductTypePartial()`). StarcounterSamples use that technique.

**Comment by miyconst** [comment]
>So it must be related to the implicit session. @miyconst it seems that you will need to tweak your idea

The initial idea was that only "Pages" will inherit from `Partial`, and by "Pages" I mean only those which will on the root level. Others should be just `Json`.

I am wrong?

**Comment by warpech** [comment]
> The initial idea was that only "Pages" will inherit from `Partial`, and by "Pages" I mean only those which will on the root level. Others should be just `Json`.

> I am wrong?

Since partials can be nested at any level, I think it is the best if no such assumption is made. Note that both `Json` at `Page` work at any level.

**Comment by per-samuelsson** [comment]
> It seems that the JSON returned to the client is not the root of the partial. See the screenshot from the debug aid:

Exactly. Thats what I tried to describe with this:

> For some reason, a partial referenced from my main page does not render, and looking at the JSON in the debug aid shows me just 3 out of like 7-8 properties.

Guess that was to unclear somehow. :smile: It did return the full result when the code behind derived `Page`. When I switched to `Partial`, this behaviour was introduced.

**Comment by per-samuelsson** [comment]
> Problem 1 can be worked around by changing the line

```
partial class ProductTypeParent : Partial, IBound<ProductType> {
``` 

to 

```
partial class ProductTypeParent : Json, IBound<ProductType> {
```

Well, OK. But then it feels I'm not really testing the `Partial` class anymore, which was what I gave feedback on. :sunglasses: Like I wrote, when I had `Page` as the base class, everthing worked fine. Just revert the last commit and test it if you like. (But you need the launcher to do so o/c).

**Comment by miyconst** [comment]
>But then it feels I'm not really testing the Partial class anymore

Fair enough. Well, I don't have a better idea how to solve this issue, but I really don't want to manually create session each time I am going to use implicit standalone mode.

**Comment by warpech** [comment]
I think that Problem 1 and Problem 2 are related to the behavior of Session, so they should be solved in Session not in Partial. Let's wait for what @chrhol has to say.

**Comment by per-samuelsson** [comment]
Any progress on this? Any consensus?

Right now, its quite painful to use `Partial` unfortunately. I find myself having to switch to `Json` or `Page` over and over because what ends up on the client with `Partial` is just..., yeah, weird. Properties missing, nothing being displayed, etc.

The sad thing about it is I really love what it tries to solve. Developing stand-alone applications is really what I'd prefer as a first thing, and not having to consider the launcher at all. Right now, there is no nice way to do so though. :cry:

**Comment by Starcounter-Jack** [comment]
I think we should get rid of ```Page``` and ```Partial``` altogether. I don't think this functionality belongs to any special base class. The Json objects should concern themselves with view models. The puppet/html functionality should not be the responsibility of the view models.

**Comment by Starcounter-Jack** [comment]
I think that using  ```<!doctype>``` as the detector is asking for trouble. In addition for it to be omitted by many authors, it is actually optional if you are using xhtml5.
